### PR TITLE
Template documentation (readme)

### PR DESCRIPTION
**Pull request by KlausLoeffelmann**
Ament

### CODE REVIEW

File: src/Resources/Custom TypeEditor Template.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
+Custom TypeEditor Template
+
+Since .NET 3.1 started to support the WinForms Runtime, a new WinForms designer
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:3]: Hard wrapping makes it harder to review.

File: src/Resources/Custom TypeEditor Template.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+was needed to support .NET applications. The work required a near-complete
+rearchitecting of the designer, as we responded to the differences between .NET
+and the .NET Framework based WinForms designer everyone knows and loves.
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:6]:
```suggestion
Windows Desktop support was (re-)introduced with .NET Core 3.1, and a new WinForms designer
was needed to support .NET applications. The work required a near-complete
rearchitecting of the designer, as we responded to the differences between .NET
and the .NET Framework based WinForms designer everyone knows and loves.
```

Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:6]: I read and re-read this sentence but I'm still having a hard time parsing it, I suggest a re-write/word.
"_The work required a near-complete rearchitecting of the designer, as we responded to the differences between .NET and the .NET Framework based WinForms designer everyone knows and loves._"

File: src/Resources/Custom TypeEditor Template.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+Until we added support for .NET Core applications there was only a single
+process, devenv.exe, that both the Visual Studio environment and the application
+being designed ran within. But .NET Framework and .NET Core can’t both run
+together within devenv.exe, and as a result we had to take the designer out of
+process, thus we called the new designer – WinForms Out of Process Designer (or
+OOP designer for short). We call the existing process – the process that Visual
+Studio runs in – the client process, and the process which shows the actual Form
+at Design time, the server process, which we call *DesignToolsServer*.
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:15]:
```suggestion
together within devenv.exe, and as a result we had to take the designer out of
process, thus we called the new designer – _WinForms out-of-process designer_. We call the existing process – the process that Visual
Studio runs in – the "client process"; and the process which shows the actual design surface (such as a form or a user control)
at design time - the "server process", which is provided by DesignToolsServer.exe.
```

IIRC, Beth Massi issued a recommendation against the use of "OOP designer".

File: src/Resources/Custom TypeEditor Template.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+# Enter the DesignToolsServer
+
+While simple Control Designer scenarios like Type Converters, Action Lists or
+CodeDom Serializers don’t need any substantial rewrites, Type Editors are a
+different beast altogether.
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:21]: Unnecessary capitalisation throughout. I'm not going to comment on every instance.
```suggestion
While simple control designer scenarios that use type converters, action lists or
CodeDom serializers may not need any substantial changes, designers that use type editors require significantly more authoring effort.
```

File: src/Resources/Custom TypeEditor Template.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+To illustrate the problem created by introducing this DesignToolsServer as a
+dedicated server process, let’s look at a typical type editor for any image
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:24]: This kind of reads like "the oil is oily".
```suggestion
To illustrate the problem created by an introduction of a
dedicated server process, let’s look at a typical type editor for any image
```

File: src/Resources/Custom TypeEditor Template.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+processing control property like the Button’s *BackgroundImage* property: While
+the actual Image that you picked will be rendered on a Button in the Server
+process, the dialog you picked it *with* runs in the context of Visual Studio.
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:27]: Type (where we refer to an _actual_ type) and properrties should formatted with backticks.


```suggestion
To illustrate the problem created by introducing this DesignToolsServer as a
dedicated server process, let’s look at a type editor for an image property like (e.g., the button’s `BackgroundImage` property).
While the actual image that you pick will be rendered in the server
process, the dialog you picked it *with* runs in the context of Visual Studio (i.e., the client process).
```

File: src/Resources/Custom TypeEditor Template.md [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+# Projects which the templates create
+
+Setting all these things up manually means coordinating a lot of moving parts,
+and there is a huge potential that things go wrong. The individual projects
+created by this template help to prevent falling into those traps. The templates
+create a series of projects and important Solution Folders, depending on your
+needs for both C\# and Visual Basic:
+
+-   **\_Solution Items:** This is a solution folder, which holds this readme,
+    the Directory.Build Target which determines the NuGet package version for
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:47]:
```suggestion
    the MSBuild props and targets which determines the NuGet package version for
```

File: src/Resources/Custom TypeEditor Template.md [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+client-side
+
+// ViewModel accesses its PropertyStore property, which in turn gets the
+required PropertyStore
+
+// proxy object directly from the server-side ViewModel.
+
+value = viewModelClient.PropertyStore;
+
+}
```
Comment (RussKie) [inline src/Resources/Custom TypeEditor Template.md:302]: Broken formatting. Code samples must be wrapped with tripple backticks followed by the language code (`cs` for C# or `vb` for Visual Basic).

**Review by RussKie — COMMENTED** [review]
Partial review of the first part of the readme. 

A lot of random capitalisation, and there some long sentences which I _personally_ found difficult to process, and I had to re-read those several times trying to grasp the idea. Please consider writing shorter sentences, as this could help making the document more approchable especially for the non-native English readers.

Also, please avoid using spaces in file names.

**Comment by KlausLoeffelmann** [comment]
Oh wait for it. It's not the final version - I am still working on that.

**Comment by kirsan31** [comment]
> Run the prepareTemplates.bat batch file.

I got this:
image

I suggest to change:
https://github.com/KlausLoeffelmann/NetControlDesigners/blob/1e857e2f379897f4bdadd874fd2f40fb74d1eb19/src/Templates/prepareTemplates.bat#L29-L31

like that:
```but
 dotnet new -u Microsoft.WinForms.Designer.TypeEditorTemplate 
 dotnet pack 
 dotnet new -i .\bin\Debug\Microsoft.WinForms.Designer.TypeEditorTemplate.1.1.0-prerelease-preview3.nupkg 
```

**Comment by KlausLoeffelmann** [comment]
> why is there a single number?

Not sure entirely what you mean, but I guess 4.7.2?
Yes, that is _a_ single number (for now), because that's what Visual Studio TFM is.
Remember, you are not providing your controls for Framework and for .NET at the same time. These templates are _only_ to provide the controls for .NET (from version 6 including onwards). The fix 4.7.2 assemblies are ONLY to run the Designer _for_ the .NET 6 control in the context of Visual Studio.

And yes, I think I'll take your suggestion, and then merge it back into my repo here (I am currently working in a Microsoft repo which is still private, but I'll ping you first as soon as we're open that up.), so you can test it.

Also, I'll continue discussing this with my boss @merriemcgaw today, and I think including the migration steps from a Framework based Type Editor makes perfect sense. I'd adapt a Framework version of this sample then, and we should include the important steps to go - with the help of the template - from Framework to .NET.

**Comment by KlausLoeffelmann** [comment]
It's not the sample project. It's something more basic a general.
The sample project was the Tile Repeater. This is just a simple Type Editor Dialog, which serves as a starting point for your own adaptation. Without a complete implementation of _a_ simple type editor, how would you approach creating a new type editor?

Also I think one Type Editor per new template should be enough. If we add another Type Editor, we would need to add I think at least 10 additional class files, and I don't think that would add clarity.

Again, I can offer you a 1:1 via teams or zoom, to take your most pressuring questions offline, if you like. I think I can quickly explain where most of the confusion comes from.

**Comment by KlausLoeffelmann** [comment]
I missed to push a commit, sorry, my bad.
Please try this, it should be better.

**Comment by kirsan31** [comment]
@KlausLoeffelmann Now I think I've got the point :)

> Again, I can offer you a 1:1 via teams or zoom, to take your most pressuring questions offline, if you like. I think I can quickly explain where most of the confusion comes from.

I think now there is definitely no need for this, I will slowly understand the projects received from the templates and try to adapt them to the chart...

At the moment the general gist is clear. I have only one question (non critical):

I totally do not understand how it ([winforms-datavisualization](https://github.com/kirsan31/winforms-datavisualization)) manages to work now being in one .net6 project? Perhaps understanding this, it will be possible to come up with some kind of temporary workaround?


Thank you once again, I hope to provide you with my feedback as I progress :)

**Comment by kirsan31** [comment]
> I am not sure I understand the question fully.

[winforms-datavisualization](https://github.com/kirsan31/winforms-datavisualization) (with custom designer and many custom editors) port is pure .Net6, one dll control. Which I compiled under the new designer SDK, slightly correcting and commenting out non-working things - that's all! And the new VS (up to 17.4) designer works with this control (if referenced by dll)! how can it be? If editors must be compiled for net472?

**Comment by KlausLoeffelmann** [comment]
VS 17.4 is still targeting ("written in") Framework 4.7.2.
And only the Type Editor part is targeting 4.7.2. 
But it's never needed at runtime.

**Comment by KlausLoeffelmann** [comment]
And the new VS (up to 17.4) designer works with this control (if referenced by dll)! how can it be?

Some things are delegated behind the scenes:
* Action Lists
* CodeDom Serializer
* Type Converters
* Some (Drop Down) Editors (which are essentially TypeEditors).

Modal-Dialog-Type-Editors won't work. If you can show them at all, they will crash or hand Visual Studio.

**Comment by KlausLoeffelmann** [comment]
There is a new version of the readme - considerably extended, also based on your question. Please re-read it. I guess it will answer most of your additional questions.

**Comment by kirsan31** [comment]
> here is a new version of the readme - considerably extended, also based on your question. Please re-read it. I guess it will answer most of your additional questions.

> Modal-Dialog-Type-Editors won't work. If you can show them at all, they will crash or hang Visual Studio.

In readme of this PR I found only this:

> Since the modern WinForms Designer runs out of the Visual Studio process for showing most of the design-time experience, type editors for **complex types** need to be migrated so that they can cooperate with the new realities of the modern WinForms Designer.

In main readme I found [this](https://github.com/KlausLoeffelmann/NetControlDesigners/blob/main/README.md#:~:text=If%20you%20just,the%20two%20processes.):

> 1. If you just have a UI, which is based on a type-converter and therefore shown in the context of the Property grid (like Enums or dedicated items to show in a property grid’s property grid cell ComboBox), your UI will be supported by the new designer model out of the box.
> 
> 2. **If you have a UI, which is part of the control (like custom painted adorners or Action Lists)**, then you would need to write your control library against the WinForms Designer SDK, but you don’t need to roundtrip data to the Server process. Everything from the Developer’s perspective seems to actually be done server-side, and you can reuse most of the existing Control Designer Code, but just need to target it against the Windows Designer SDK.
> 
> 3. If you have custom type editors, however, which are displaying dedicated modal dialogs, then there is some rewriting effort involved for roundtripping the required data between the two processes.

1. is ok. In 2. I think bolded part need a bit more clarification.. And 3. is about modal dialogs again...

So we make conclusion that complex types are types that need modal dialogs in editing...

These dialogs are definably modal:
image

And they've been working great (**without any crash or even error**) for almost a year now...
There is clearly some kind of misunderstanding / understatement here. 🤔

**Comment by KlausLoeffelmann** [comment]
Did the `CallOffer` or `DataPoint` classes/structs have EditorAttributes in the Framework version - so, do they bring their own Type Editors? Or did the Editors work there intrinsically?

**Comment by KlausLoeffelmann** [comment]
Did you compare the behavior of the Collection Editors from Framework and the last working version of .NET? My suspicion here is that the custom editors did not get pulled it. Rather, the out-of-process Designer tried to provide its own collection editor functionality as well as it was able to. But then it may hit a spot, where customization was needed. It could have been pure luck, that it was somehow working before, but also probably never the way it was intended to work.

I have currently no time currently to investigate that thoroughly, but maybe I can manage to take a quick look later this week.

You would need to figure out, why custom type editors were needed to begin with. What are they doing that the intrinsic editors cannot do? If we need to implement a custom collection type editor for each type, that would be even more effort, unfortunately. There is also a sample ("Sophisticated") in this repo, showing, how custom collection editors are implemented. In a nutshell, you would need to provide additional functionality via Endpoints for the Collection operations like Add, Delete, Edit... .

But, to be honest, there is a lot of assuming here. Let's take the time to get behind it.

Also, Collection Editors will not be covered in the first part of the blog post. We need to get this basic information out first, and then we aim with the second part in the first quarter of next year. The reason is, we're thinking about steps to easing up the necessary boiler plate coding, especially for Collection Editors.

**Comment by KlausLoeffelmann** [comment]
I took a quick look:
Setting up a Chart-Series, for example, is a comparatively complex functionality in the old Framework editor. There is a dedicated `SeriesCollectionEditor` for that. I am _pretty_ sure that that wasn't the Editor you saw in the working version.

**Comment by kirsan31** [comment]
> I took a quick look: Setting up a Chart-Series, for example, is a comparatively complex functionality in the old Framework editor. There is a dedicated `SeriesCollectionEditor` for that. I am _pretty_ sure that that wasn't the Editor you saw in the working version.

.Net472:
Snipaste_2022-10-26_20-05-11
.Net6
Snipaste_2022-10-26_20-07-00

The only difference is that some custom properties don't have a color/picture icon. From here it follows that the [GradientEditor](https://github.com/kirsan31/winforms-datavisualization/blob/6f35dccb327f78f889797ce9d610764199edecab/src/System.Windows.Forms.DataVisualization/Design/GradientEditor.cs#L25) does not work in .Net6. This is also a good question - why (in my understanding it's p.1 may be p.2)? And of course why almost all of `CollectionEditor` inheritors - work???

**Comment by KlausLoeffelmann** [comment]
OK, give me quick pointer, so I don't need to browse all the threads:
Which version of Visual Studio/the WinForms Designer does work again, and from which on not anymore. 

And also, which branch from which repo again should I get?

**Comment by KlausLoeffelmann** [comment]
Never mind, I got it.

**Comment by kirsan31** [comment]
I'll post it anyway so it's in one place.
Repo to test: https://github.com/kirsan31/winforms-datavisualization/tree/dev
Draft _blind_ port to new designer commit: https://github.com/kirsan31/winforms-datavisualization/commit/17698a2255af0eb504588a6ee58d83cc36394924
Designer was working in all versions of V.S. from 17.0 to 17.4 Preview 1. And [was broken in 17.4 Preview 2](https://github.com/kirsan31/winforms-datavisualization/issues/4).
Also this [very strange serialization bug](https://github.com/kirsan31/winforms-datavisualization/issues/5) (Toward the end of the post, the reason (line of code) causing this bug is described) was always there.

**Comment by kirsan31** [comment]
@KlausLoeffelmann 
Why we are waiting on https://github.com/kirsan31/winforms-datavisualization/issues/4 and https://github.com/kirsan31/winforms-datavisualization/issues/5 I will try to formulate the main questions / difficulties with the new designer.

The **main** problem is that we have no strict description what should be done and why with custom editors. All we have is [this notes](https://github.com/KlausLoeffelmann/NetControlDesigners/blob/main/README.md#:~:text=but%20only%20if,the%20two%20processes.):

> but only if there is an actual UI which needs to be shown on top of the UI that is presented in the context of your actual control. Here is what that means exactly:
> 
> - If you just have a UI, which is based on a type-converter and therefore shown in the context of the Property grid (like Enums or dedicated items to show in a property grid’s property grid cell ComboBox), your UI will be supported by the new designer model out of the box.
> 
> - If you have a UI, which is part of the control (like custom painted adorners or Action Lists), then you would need to write your control library against the WinForms Designer SDK, but you don’t need to roundtrip data to the Server process. Everything from the Developer’s perspective seems to actually be done server-side, and you can reuse most of the existing Control Designer Code, but just need to target it against the Windows Designer SDK.
> 
> - If you have custom type editors, however, which are displaying dedicated modal dialogs, then there is some rewriting effort involved for roundtripping the required data between the two processes.

From which I conclude that the _hard_ path of cross-process interaction is only needed for editors with modal dialogs. But this conclusion is fundamentally wrong. Using the chart control as an example, we make sure that:
1. _Default_ (`Microsoft.DotNet.DesignTools.Editors.CollectionEditor`) modal dialogs can be displayed without cross-process.
2.  None of the **custom** editors work at all! At first glance, they work, but in reality no modification has been applied - all editors use the default implementations from the base classes. **No overrides calls at all**, tested on editors derived from:
`System.Drawing.Design.UITypeEditor` 
`Microsoft.DotNet.DesignTools.Editors.CollectionEditor`

Is p.2 a bug or is it intended? If the latter, then it turns out that if you need **even the slightest change** in the default behavior, then you are in trouble 😭

**Pull request closed without merging**